staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -0,0 +1,270 @@
+		}
+	}
+}
+
+func (rdm *resourceDiscoveryManager) AddGroups(groups ...apidiscoveryv2beta1.APIGroupDiscovery) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	for _, group := range groups {
+		for _, version := range group.Versions {
+			rdm.addGroupVersionLocked(group.Name, version)
+		}
+	}
+}
+
+func (rdm *resourceDiscoveryManager) AddGroupVersion(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.addGroupVersionLocked(groupName, value)
+}
+
+func (rdm *resourceDiscoveryManager) addGroupVersionLocked(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery) {
+	klog.Infof("Adding GroupVersion %s %s to ResourceManager", groupName, value.Version)
+
+	if rdm.apiGroups == nil {
+		rdm.apiGroups = make(map[string]*apidiscoveryv2beta1.APIGroupDiscovery)
+	}
+
+	if existing, groupExists := rdm.apiGroups[groupName]; groupExists {
+		// If this version already exists, replace it
+		versionExists := false
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
+			existing.Versions = append(existing.Versions, value)
+		}
+
+	} else {
+		group := &apidiscoveryv2beta1.APIGroupDiscovery{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "APIGroupDiscovery",
+				APIVersion: "apidiscovery.k8s.io/v2beta1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []apidiscoveryv2beta1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroups[groupName] = group
+		rdm.apiGroupNames = append(rdm.apiGroupNames, group.Name)
+
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			break
+		}
+	}
+
+	if len(group.Versions) == 0 {
+		for i, g := range rdm.apiGroupNames {
+			if g == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			}
+		}
+		delete(rdm.apiGroups, group.Name)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	rdm.lock.RLock()
```

Comment:
I'm confused with the lock dance in this function, why can't we just keep the lock the whole time?

Reply:
It is an RWMutex. Could keep it as a `Lock` the entire time, but the would mean multiple HTTP requests could not be served at a time in the average case

Reply:
afaik this is also addressing David's comment: https://github.com/kubernetes/kubernetes/pull/112113#discussion_r967399061.

Reply:
There are a lot of locks here.  I think part of my hesitation with this factoring is that there are two completely different sets of responsibilities in this feature:

1. Generate new aggregated discovery doc based on incoming changes
2. Serve the newest aggregated doc if available, the last one and Stale, or block/return error if no doc available yet

Mixing the responsibilities for these in one type are:

a. creating higher contention than necessary (1 above could be doing a lock and swap, not lock and calculate)
b. generating a very complicated ServeHTTP method that is hard to update later

I would like to see those responsibilities split and testable independently, not merged.

Reply:
I read 3 different mostly orthogonal points:
1. How can we simplify the locking situation here
2. Should we organize the code with more SOLID principles so that they can be tested independently
3. Should computations (building the list, calculating the etag) be done lazily or pro-actively

I agree that 1 looks complicated. We could acquire the write lock and demote it if `rmd.groups` is non-nil/after we've updated `rdm.groups` which would make things much easier to read, though it would also aquire the write lock for no reason in many cases.

I agree with 2., we could test that code separately, but it's also just creating a list.

Not sure about 3., I don't think any of this is complicated enough that we should have an extra-thread (which, in turn will definitely make things more complicated).

Reply:
Also, I agree that we could probably put `rdm.groups` behind an atomic pointer to load and swap without locking (we could probably write this entire serverHTTP function without any write lock), but I don't know if that's necessary or useful here specifically. The write locked section is fairly small and probably not super contentious anyway?

Reply:
Rewrote the ServeHTTP function to avoid using a single lock. I think it looks a lot cleaner with an atomic lock, still working on a couple of tests.